bigquery_etl/public_data/publish_gcs_metadata.py

```diff
@@ -0,0 +1,202 @@
+    match = GCS_FILE_PATH_RE.match(gcs_path)
+
+    if match is not None:
+        return (match.group("dataset"), match.group("table"), match.group("version"))
+    else:
+        return None
+
+
+def get_public_gcs_table_metadata(
+    storage_client, bucket, api_version, endpoint, target_dir
+):
+    """Return a list of metadata of public tables and their locations on GCS."""
+    prefix = f"api/{api_version}"
+
+    blobs = storage_client.list_blobs(bucket, prefix=prefix)
+    blob_paths = [blob.name for blob in blobs]
+
+    return [
+        GcsTableMetadata(list(files), endpoint, target_dir)
+        for table, files in groupby(blob_paths, dataset_table_version_from_gcs_path)
+        if table is not None
+    ]
+
+
+def publish_all_datasets_metadata(table_metadata, output_file):
+    """Write metadata about all available public datasets to GCS."""
+    metadata_json = {}
+
+    for metadata in table_metadata:
+        if metadata.dataset not in metadata_json:
+            metadata_json[metadata.dataset] = {}
+
+        dataset = metadata_json[metadata.dataset]
+
+        if metadata.table not in dataset:
+            dataset[metadata.table] = {}
+
+        table = dataset[metadata.table]
+
+        if metadata.version not in table:
+            table[metadata.version] = {}
+
+        table[metadata.version] = metadata.table_metadata_to_json()
+
+    logging.info(f"Write metadata to {output_file}")
+
+    with smart_open.open(output_file, "w") as fout:
+        fout.write(json.dumps(metadata_json))
+
+
+def publish_table_metadata(table_metadata, bucket):
+    """Write metadata for each public table to GCS."""
+    for metadata in table_metadata:
+        output_file = f"gs://{bucket}/{metadata.files_path}/metadata.json"
+
+        logging.info(f"Write metadata to {output_file}")
+        with smart_open.open(output_file, "w") as fout:
+            fout.write(json.dumps(metadata.files_metadata_to_json()))
+
+
+def main():
+    """Generate and upload GCS metadata."""
+    args = parser.parse_args()
+    storage_client = storage.Client(args.project_id)
+
+    # set log level
+    try:
+        logging.basicConfig(level=args.log_level, format="%(levelname)s %(message)s")
+    except ValueError as e:
+        parser.error(f"argument --log-level: {e}")
+
+    if os.path.isdir(args.target):
+        gcs_table_metadata = get_public_gcs_table_metadata(
+            storage_client,
+            args.target_bucket,
+            args.api_version,
+            args.endpoint,
+            args.target,
+        )
+
+        output_file = f"gs://{args.target_bucket}/all_datasets.json"
+        publish_all_datasets_metadata(gcs_table_metadata, output_file)
+        publish_table_metadata(gcs_table_metadata, args.target_bucket)
+    else:
+        print(
+            """
+            Invalid target: {}, target must be a directory with
+            structure /<dataset>/<table>/metadata.yaml.
+            """.format(
```

Comment:
This multiline string will have be printed with indentation we don't want. We may want to change this to several single line string literals.